packages/report-server/src/reportBuilder/query/buildOrganisationUnitParams.ts

```diff
@@ -0,0 +1,47 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2021 Beyond Essential Systems Pty Ltd
+ */
+
+import { ReqContext } from '../context';
+import { FetchReportQuery, ReportConfig } from '../../types';
+
+const REQUESTED_ORG_UNITS_PLACEHOLDER = '$requested';
+
+export const buildOrganisationUnitParams = async (
+  ctx: ReqContext,
+  query: FetchReportQuery,
+  config: ReportConfig,
+) => {
+  const { hierarchy, accessPolicy, permissionGroup } = ctx;
+  const { organisationUnitCodes: requestedCodes } = query;
+  const { organisationUnits: organisationUnitsSpec } = config.fetch;
+
+  if (!organisationUnitsSpec) return requestedCodes;
+
+  const codes: string[] = [];
+  const codesToFetch: string[] = []; // must fetch codes from entityApi to ensure we have permissions
+  organisationUnitsSpec.forEach(organisationUnit => {
+    if (organisationUnit === REQUESTED_ORG_UNITS_PLACEHOLDER) {
+      codes.push(...requestedCodes);
+    } else {
+      codesToFetch.push(organisationUnit);
+    }
+  });
+
+  const fetchedEntities = await ctx.services.entity.getEntities(hierarchy, codesToFetch, {
+    fields: ['code', 'country_code', 'type'],
+  });
+
+  const fetchedCodesWithAccess = fetchedEntities
+    .filter(({ type }) => type !== 'project') // Can't fetch at project level due to permissions, not supported for now
```

Comment:
So this is a bit annoying. We strictly speaking don't support fetching data at the project level, as we have no way of enforcing permissions on it. Instead if such a request is passed to report-server atm, we map the project code to the underlying countries. I'm not the biggest fan of that behaviour and it can lead to some unexpected things, so I chose to not do that here... Perhaps it would be worth it tho... not sure...

Reply:
I think it's best to just do an easy unsupported check like you've done rather than wrangling it to awkwardly work, and if we need it to work later then there's just a bunch of refactor in different places to do it properly

In this case I think the `country_code !== null` is doing enough work for us though. There's one case in my local db where it's a project with a country code set, and that's a Fiji specific project, so that seems to give us the opportunity to opt-in to fetching data against a project if it's just for one country